tensorflow-framework/src/test/java/org/tensorflow/framework/metrics/AUCTest.java

```diff
@@ -0,0 +1,326 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.framework.metrics;
+
+import org.junit.jupiter.api.Test;
+import org.tensorflow.Operand;
+import org.tensorflow.framework.utils.TestSession;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Ops;
+import org.tensorflow.types.TFloat32;
+import org.tensorflow.types.TInt32;
+import org.tensorflow.types.TInt64;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.tensorflow.framework.utils.CastHelper.cast;
+
+public class AUCTest {
+  private final TestSession.Mode tfMode = TestSession.Mode.GRAPH;
+
+  float epsilon = 1e-4F;
+
+  int numThresholds = 3;
+  float[] predArray = new float[] {0f, 0.5f, 0.3f, 0.9f};
+  int[] trueArray = new int[] {0, 0, 1, 1};
+  float[] sampleWeight = new float[] {1, 2, 3, 4};
+
+  @Test
+  public void testValueIsIdempotent() {
+    try (TestSession session = TestSession.createTestSession(tfMode)) {
+      Ops tf = session.getTF();
+      Operand<TFloat32> yPred = tf.constant(predArray);
+      Operand<TInt32> yTrue = tf.constant(trueArray);
+      AUC<TFloat32> instance = new AUC<>(tf, numThresholds, 1001L, TFloat32.class);
+
+      session.run(tf.init());
+
+      Op update = instance.updateState(yTrue, yPred, null);
+
+      for (int i = 0; i < 10; i++) {
+        session.run(update);
```

Comment:
What effect do we expect from these repeated calls to `Session.run`? Does it exactly mirror a use of our API that we expect in production? For example, does every iteration force a fresh execution of the `Op` returned by `updateState`, or does it merely retrieve a value cached from the first execution? Is it the same as, or different from, a caller in production repeatedly forming control dependencies on the `Op` returned by `updateState`?

Reply:
It re-executes the updateState operation.  This particular test tests that the results do not change on each invocation.

In some other Class Tests, the results do change, especially if the updateState operation is doing something like using Random number generation. Sometimes this kind of behavior is buried deep in the TF raw Operations.